### PR TITLE
Sort peek results in the replica instead of the coordinator

### PR DESCRIPTION
**Pull request by sdht0**
Skunkworks while planning a wedding. Continued thinking about the project and came up with this solution. The tricky part was to get the merge logic represented in a way that satisfies the existing constraints, e.g., should be `Sync+Send`, hence no references (?).

### Motivation

Currently, peek results are sorted in the coordinator. For `k` workers each producing `N` results, the sorting cost is `(N*k)*log(N*k)` in the single coordinator thread.

This PR shifts the sorting to the replicas when possible, leaving just a `k`-way merge in the coordinator. The new cost is `(N*k)*(log(N)+log(k))`, for sorting `N` results at each of the `k` workers and a final merge of `N*k` results at the coordinator using a binary heap. Moreover, the bulk of the computation occurs across multiple workers instead of a single coordinator thread.

**Note that fast peeks and other results that are not computed in the replica will still be sorted in the coordinator.**

### Tips for reviewer

* I haven't had a chance to run benchmarks yet. Will take a stab at it after CI passes.

cc @teskje 

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

File: src/expr/src/relation.rs [hunk 1]
```diff
@@ -3541,31 +3541,11 @@ impl RowSetFinishing {
     /// Implementation for [`RowSetFinishing::finish`].
     fn finish_inner(
         &self,
-        rows: RowCollection,
-        max_result_size: u64,
+        rows: Vec<RowCollection>,
+        _max_result_size: u64,
```
Comment (sdht0) [inline src/expr/src/relation.rs:3545]: Do we still need to check for `max_result_size`?

File: src/repr/src/row/collection.rs [hunk 2]
```diff
@@ -511,37 +629,36 @@ mod tests {
...
+            .into_row_iter()
+            .apply_offset(1);
         assert_eq!(iter.next(), Some(b.borrow()));
         assert_eq!(iter.next(), Some(a.borrow()));
         assert_eq!(iter.next(), Some(a.borrow()));
         assert_eq!(iter.next(), Some(a.borrow()));
         assert_eq!(iter.next(), None);
         assert_eq!(iter.next(), None);
 
-        let col = iter.into_inner();
```
Comment (sdht0) [inline src/repr/src/row/collection.rs:529]: Removed this function as the collection removes entries at the end of an iteration.

To keep the logic simple, simply recreating the iterator from scratch every time.

File: src/repr/src/row/collection.rs [hunk 3]
```diff
@@ -225,32 +267,143 @@ impl RustType<ProtoEncodedRowMetadata> for EncodedRowMetadata {
...
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for HeapRowCollection {
+    fn cmp(&self, other: &Self) -> Ordering {
+        // The comparison is done the other way to ensure a _min-heap_ when inserting into a
+        // [`BinaryHeap`], which is a _max-heap_ by default.
+        other.get_row().cmp(&self.get_row())
```
Comment (antiguru) [inline src/repr/src/row/collection.rs:310]: This uses the row sort order, which isn't the same as the order specified by the `ORDER BY` clause. We need to use the row set finishing here, and compare the rows based on that order.

Comment (sdht0) [reply src/repr/src/row/collection.rs]: Thanks, fixed.

File: src/compute/src/compute_state.rs [hunk 4]
```diff
@@ -1535,7 +1536,11 @@ impl IndexPeek {
             }
         }
 
-        Ok(results)
+        // Sort results. The coordinator collects sorted results from all workers and performs
+        // an N-way merge to produce the final results. This way most of the sorting happens in
+        // the workers, as compared to merging unsorted results and sorting in the coordinator thread.
+        results.sort_by(sort_by);
```
Comment (sdht0) [inline src/compute/src/compute_state.rs:1542]: Is this unconditional sort correct? Specifically, can we rely on the `sort_by` closure above to correctly sort the rows whether an `order_by` clause is present or not?

File: src/repr/src/row/collection.rs [hunk 5]
```diff
@@ -131,42 +155,39 @@ impl RowCollection {
...
-        for row in rows {
-            encoded.extend(row.data());
-            metadata.push(EncodedRowMetadata {
-                offset: encoded.len(),
-                diff: unsafe { NonZeroUsize::new_unchecked(1) },
-            });
-        }
+    pub fn sorted_view_cost(&self) -> usize {
+        let sorted_view_mem = self.entries().saturating_mul(size_of::<usize>());
+        self.byte_len().saturating_add(sorted_view_mem)
```
Comment (sdht0) [inline src/repr/src/row/collection.rs:186]: Shifted from `relation.rs`.

Why is `self.byte_len()` added to the cost? Isn't the existing allocation used when creating the sorted view?

File: src/repr/src/row/collection.rs [hunk 6]
```diff
@@ -225,32 +332,163 @@ impl RustType<ProtoEncodedRowMetadata> for EncodedRowMetadata {
...
+#[derive(Clone)]
+pub struct HeapRowCollection {
+    /// A sorted [`RowCollection`].
+    row_collection: SortedRowCollection,
+    /// The `row_collection` index that is considered active in the heap.
+    row_index: usize,
+    /// The number of diffs left to process of the current row.
+    diffs_left: usize,
+    /// Sort order
+    sort_by: Arc<Mutex<dyn FnMut(&RowRef, &RowRef) -> Ordering + Sync + Send>>,
```
Comment (sdht0) [inline src/repr/src/row/collection.rs:349]: The `Mutex` here allows interior mutability but doesn't really feel right. The tradeoffs are:
* We need a `FnMut` here to reuse `Vec` allocations across all comparisons.
* std::BinaryHeap (needed for a `N*log(k)` merge) needs an `Ord` comparison which does not have mutable access.
* Any type here needs to be `Send + Sync`, hence no `RefCell` etc.

Perhaps I'm missing a way to simplify this logic.

**Review by sdht0 — COMMENTED** [review]
Turns out that of course not all results come from the replica (e.g., fast peeks) and thus we need to check if row collections are actually sorted and sort them if not. Introduced a new `RowDiffs` type that tracks this information.

`sqllogictests` are passing locally for me. Let's see if CI catches anything else, once it's unblocked from running.

File: src/expr/src/relation.rs (hunk 1 above)
Comment (sdht0) [reply src/expr/src/relation.rs]: Yes, we do.

**Comment by sdht0** [comment]
Pinging to get some eyes on this.

cc @teskje

**Comment by teskje** [comment]
> Turns out that of course not all results come from the replica (e.g., fast peeks) and thus we need to check if row collections are actually sorted and sort them if not.

Hm, I think there is a misunderstanding here. `PeekResponse::Row` instances are only created by replica code (apart from the merging done in `PartitionedComputeState`). That's true for fast-path peeks as well. The only difference between slow-path and fast-path peeks is that the former install a temporary dataflow while the latter use one that's already installed.

**Comment by sdht0** [comment]
Thanks for taking a look!

> That's true for fast-path peeks as well. The only difference between slow-path and fast-path peeks is that the former install a temporary dataflow while the latter use one that's already installed.

I should have said [PeekPlan::FastPath](https://github.com/MaterializeInc/materialize/blob/bbebf317e6f6e84553f6c11593d265a7563d03ae/src/adapter/src/coord/peek.rs#L459). Selects on constant values do not go to the replica and are still expected to be sorted at the coordinator, such as [this test](https://github.com/MaterializeInc/materialize/blob/bbebf317e6f6e84553f6c11593d265a7563d03ae/test/sqllogictest/arithmetic.slt#L836-L854).

Row collections are also created in  `Coordinator::send_diffs()`, called from `Coordinator::sequence_read_then_write` and `Coordinator::insert_constant`. Not sure about the pre-sorting of rows here.

> PeekResponse::Row instances are only created by replica code (apart from the merging done in PartitionedComputeState).

The other place is `PersistPeek::do_peek()`. But these results also always come from a dataflow and will be always sorted?

Since it is now critical for correctness that Rows should be sorted, the `RowDiffs::sorted()` method captures this requirement explicitly, especially if other ways to construct rows are added in the future.

**Comment by bkirwi** [comment]
> The other place is PersistPeek::do_peek(). But these results also always come from a dataflow and will be always sorted?

These don't come from a dataflow, though they do come from clusterd... and they are guaranteed to satisfy the requested sort. (Today we do this by just not generating Persist peeks when a sort is requested, but we plan to relax that soon.)j

**Comment by bkirwi** [comment]
> To confirm, when a sort is not requested (which I assume means a missing `ORDER BY`), is the [default sort](https://github.com/MaterializeInc/materialize/pull/30388/files#diff-90705c3b3bdb14ef2b5daf7b2ed6b45863f0570d5ae2ce0fce4dfad9a72b4b51R1404) applied?

> In general, is it then safe to assume that `Row`s will always be sorted in `PersistPeek::do_peek()` [here](https://github.com/MaterializeInc/materialize/pull/30388/files#diff-90705c3b3bdb14ef2b5daf7b2ed6b45863f0570d5ae2ce0fce4dfad9a72b4b51R1251)?

No to both of these! All I meant to suggest is that we don't generate Persist-backed peeks that violate the `ORDER BY` - otherwise Persist almost certainly will choose a different ordering than the len / bytewise `cmp` implementation for `Row`.

I guess the other thing to note is that we only hit this codepath right now for very small result sets (up to 25 rows, I think?) so it's almost certainly fine to do the paranoid / conservative thing on this path to start.

**Pull request closed without merging**